src/lazy/preload/index.ts

```diff
@@ -10,74 +10,136 @@ import { insertLinkTag } from './utils';
+  then: () => fakePromise,
+  catch: () => fakePromise,
+  finally: () => fakePromise,
+};
+
+export function createWebpackPreloadStrategy({
+  loader,
+  rel,
+}: Pick<PreloadStrategyOptions, 'loader' | 'rel'>): PreloadStrategy {
   if (
     typeof __webpack_require__ === 'undefined' ||
     typeof __webpack_get_script_filename__ === 'undefined'
   )
-    return false;
+    throw new Error('Unsupported preload strategy');
+
+  return () => {
+    // Replace requireEnsure to create link tags instead of scripts
+    const requireEnsure = __webpack_require__.e;
+    const cleanupLinkTags: Cleanup[] = [];
+    __webpack_require__.e = function requirePreload(chunkId: string) {
+      const href = __webpack_get_script_filename__(chunkId);
+      cleanupLinkTags.push(insertLinkTag(href, rel));
+
+      return fakePromise;
+    };
 
-  // replace requireEnsure to create link tags instead of scripts
-  const requireEnsure = __webpack_require__.e;
-  __webpack_require__.e = function requirePreload(chunkId: string) {
-    const href = __webpack_get_script_filename__(chunkId);
-    insertLinkTag(href, rel);
+    try {
+      loader();
+    } catch (err) {
+      // Ignore any errors
+    }
 
-    return fakePromise;
+    // Restore real webpack require ensure
+    __webpack_require__.e = requireEnsure;
+
+    return () => {
+      for (const cleanupLinkTag of cleanupLinkTags) {
+        cleanupLinkTag();
+      }
+    };
   };
+}
 
-  try {
+export function createLoaderPreloadStrategy({
+  loader,
+}: Pick<PreloadStrategyOptions, 'loader'>): PreloadStrategy {
+  return () => {
     loader();
-  } catch (err) {
-    // ignore
-  }
-  // restore real webpack require ensure
-  __webpack_require__.e = requireEnsure;
 
-  return true;
+    return () => {
+      // Nothing to cleanup...
+    };
+  };
 }
 
-export function preloadAssetViaLoader(loader: Loader<unknown>) {
-  loader();
-
-  return true;
+function isPresent<T>(t: T | undefined | null | void): t is T {
+  return t !== undefined && t !== null;
 }
 
-export function preloadAsset(
-  loader: Loader<unknown>,
-  { moduleId, priority }: { moduleId: string; priority?: PreloadPriority }
-) {
-  if (isNodeEnvironment()) return;
+export type PreloadAssetOptions = {
+  loader: Loader<unknown>;
+  moduleId: string;
+  priority?: PreloadPriority;
+};
+
+export function preloadAsset({
+  loader,
+  moduleId,
+  priority,
+}: PreloadAssetOptions): Cleanup {
+  if (isNodeEnvironment())
+    return () => {
+      // Nothing to cleanup...
+    };
+
   const rel = priority === PRIORITY.HIGH ? 'preload' : 'prefetch';
-  [
-    preloadAssetViaManifest,
-    preloadAssetViaWebpack,
-    preloadAssetViaLoader,
-  ].some(strategy => strategy(loader, { moduleId, rel }));
+  const preloadStrategies = [
+    createManifestPreloadStrategy,
+    createWebpackPreloadStrategy,
+    createLoaderPreloadStrategy,
+  ]
+    .map(strategyFactory => {
+      try {
+        return strategyFactory({ loader, moduleId, rel });
```

Comment:
🤔 Is this working as the previous implementation? I think here we are actually kicking off all preload strategies (if available) instead of just the first one that works

Reply:
Nevermind, got confused by the triple level of function return 😆 
Still find overkill to "prepare" up to 3 strategies and then running only one... Why not bailing out earlier?

Reply:
I wanted to use native array methods, similar to the initial `.some` implementation, like `.find` but that could only be done with the nested functions. It seems that just converting it to a for loop is quite trivial, and will simplify the logic as you prefer.